### ISSUE TITLE
Cookie signature not valid on refreshing cookie

### ISSUE THREAD

**Issue by kereis** [oauth2-proxy/oauth2-proxy#1948]
## Expected Behavior

As long as I don't misunderstand `--cookie-refresh` and when using the `oidc` provider, I'd expect the proxy to refresh the access token using the refresh token and update the cookie accordingly, keeping the current session and not forcing the user to refresh the page.

## Current Behavior

When a cookie refresh happens, the proxy logs this error message:
```
[stored_session.go:94] Error loading cookied session: cookie signature not valid, removing session
```
This forces the user to refresh their page as each request on a single page application will be redirected to the identity provider of choice.

## Steps to Reproduce (for bugs)

1. Deploy a container that you want to secure via oauth2-proxy v7.4.0
2. Deploy an identity provider, like Keycloak
3. Deploy an oauth2-proxy container using these config parameters:
```yaml
# OIDC
- --provider=oidc
- --scope=openid
- --email-domain=*
# Link to e. g. Keycloak
- --oidc-issuer-url=http://localhost:8080/realms/myrealm
- --skip-provider-button=true

# Cookie settings
- --cookie-secure=true
# Assuming the access token length to be 5 minutes, we should refresh the token/cookie as soon as possible
- --cookie-refresh=2m
# A session can only last 30 minutes max. 
- --cookie-expire=30m
- --cookie-samesite=lax

- --http-address=0.0.0.0:4180
# Your container you'd want to secure
- --upstream=https://localhost:8090/

- --pass-access-token=true
- --set-authorization-header=true
- --set-xauthrequest=true
- --pass-authorization-header=true
- --skip-jwt-bearer-tokens=true

# Logging
- --request-logging=true
- --auth-logging=true
- --standard-logging=true
```

4. Create a user in your identity provider
5. When accessing your single page application, log in and keep clicking around for about 2 minutes. Make sure your interactions trigger any requests against the backend.
6. While you interact with the SPA, check the oauth2-proxy's logs.
7. After 2 minutes, you might receive frontend errors due to the request being redirect to Keycloak (302).

## Context

I'd like to secure a single page application being served by a Java backend. Everything is deployed on Kubernetes.
The ingress settings for the oauth2-proxy have already been adjusted regarding the cookie's size and nginx's buffer size.
Running the oauth2-proxy with Redis is currently not an option.
Options for passing the access token to the upstream are also set as I want to read the access token in the backend in the future.

Unfortunately, I cannot make up what this error message could actually mean. I tried to swap out cookie secrets, paid attention to its size (16, 32, 64 bit).
However, as soon as you refresh the page, you're logged in again or you need to log in (depending if your session is expired).
This behaviour deteriorates user experience in the frontend significantly, making it unpleasent to use, unless the cookie refresh feature is turned off.

Maybe anyone can point me into the right direction. 

## Your Environment

- Deployed on Kubernetes
- nginx used as ingress controller
- Both the app and oauth2-proxy run on the same domain. oauth2-proxy has its own ingress, using `/oauth2` with `pathType: Prefix`
- oauth2-proxy version used: v7.4.0

**Comment by JoelSpeed** [comment]
These kinds of issues are very hard to dig into and very hard to reproduce, the most likely way this issue will get solved is if you can add some debug logging and try and work out exactly where in the code the issue is coming from.

Your expectations around the behaviour are correct, there must be something weird going on with your `--cookie-secret` value.

One thing I might suggest you try, is to set up redis and use that as a session storage, that would help alleviate any issues with cookie splitting that might be causing problems here

**Comment by kereis** [comment]
Thank you for your response and sorry for the late response. I have to postpone this research until further notice. I can try to take a look at it if I find time to do that.

My current workaround is deactivating the cookie refresh mechanism and adjust session times for now.

**Comment by attiss** [comment]
+1 on this issue

---

I'm using an OIDC provider that supports refresh tokens and using cookie session storage.
(The size of the session cookie would exceed the 4000 bytes, therefore it's getting splitted into two.)

When the session cookie is initially created, everything works fine.

```
172.30.12.69:3573 - ba1451sdfc61o3f9fdec1239c0fc7ec4 - - [2023/03/06 16:35:25] example.com GET - "/oauth2/start?rd=%2F" HTTP/1.1 "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/110.0.0.0 Safari/537.36" 302 469 0.000
172.30.12.69:3573 - 8bf5247fasdfb1d8a16d431a8d078ce4 - user@example.com [2023/03/06 16:35:43] [AuthSuccess] Authenticated via OAuth2: Session{email:user@example.com user:xxxxxxxx PreferredUsername: token:true id_token:true created:2023-03-06 16:35:43.423069994 +0000 UTC m=+78.442413071 expires:2023-03-06 16:40:43.347606521 +0000 UTC m=+378.366949718 refresh_token:true}
[2023/03/06 16:35:43] [session_store.go:163] WARNING: Multiple cookies are required for this session as it exceeds the 4kb cookie limit. Please use server side session storage (eg. Redis) instead.
172.30.12.69:3573 - 8bf5247f6cf1asdfa16d431a8d078ce4 - - [2023/03/06 16:35:43] example.com GET - "/oauth2/callback?code=xxx" HTTP/1.1 "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/110.0.0.0 Safari/537.36" 302 24 0.180
172.30.12.69:32153 - a2700ed0fa3a9a1af40c52a21a2c828e - user@example.com [2023/03/06 16:35:43] oauth2.default.svc.cluster.local GET - "/oauth2/auth" HTTP/1.1 "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/110.0.0.0 Safari/537.36" 202 0 0.000
172.30.12.69:11587 - a468ecf0139401981asdfcd8f9d7051a - user@example.com [2023/03/06 16:35:56] oauth2.default.svc.cluster.local GET - "/oauth2/auth" HTTP/1.1 "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/110.0.0.0 Safari/537.36" 202 0 0.000
172.30.12.69:39947 - 5ed0d3dcbdfefb59d1asdf3697b21120 - user@example.com [2023/03/06 16:36:00] oauth2.default.svc.cluster.local GET - "/oauth2/auth" HTTP/1.1 "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/110.0.0.0 Safari/537.36" 202 0 0.000
172.30.12.69:7175 - e2afd4acd7885asda9234b5c77d84a66 - user@example.com [2023/03/06 16:36:12] oauth2.default.svc.cluster.local GET - "/oauth2/auth" HTTP/1.1 "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/110.0.0.0 Safari/537.36" 202 0 0.000
```

Then once the session is old enough, the refresh process is executed and the session cookies are updated:

```
[2023/03/06 16:36:14] [stored_session.go:189] Refreshing session - User: xxxxxxxx; SessionAge: 30.576930006s
[2023/03/06 16:36:15] [session_store.go:163] WARNING: Multiple cookies are required for this session as it exceeds the 4kb cookie limit. Please use server side session storage (eg. Redis) instead.
172.30.12.69:7183 - 53811ce7eaasdf2034f57598ed1ed588 - user@example.com [2023/03/06 16:36:14] oauth2.default.svc.cluster.local GET - "/oauth2/auth" HTTP/1.1 "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/110.0.0.0 Safari/537.36" 202 0 0.252
```

But on the next request, the session cookie validation fails:

```
[2023/03/06 16:36:20] [stored_session.go:94] Error loading cookied session: cookie signature not valid, removing session
172.30.12.69:62865 - 505e5addeba3d94d1232bdcd7f900f88 - - [2023/03/06 16:36:20] oauth2.default.svc.cluster.local GET - "/oauth2/auth" HTTP/1.1 "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/110.0.0.0 Safari/537.36" 401 13 0.000
172.30.12.69:3573 - 81b144afdasffgf1bd1a123422dbc944 - - [2023/03/06 16:36:20] example.com GET - "/oauth2/start?rd=%2F" HTTP/1.1 "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/110.0.0.0 Safari/537.36" 302 469 0.000
```

---

We'll try investigating the issue and provide more details in the upcoming days/weeks.

**Comment by marcsello** [comment]
Follow up on the above comment by @attiss

We have been using oauth2-proxy in a Kubernetes environment. With NGINX being the ingress controller and using the `auth_request` directive for oauth2-proxy.

Our issue was caused by the NGINX reverse proxy, which forwarded only one `set-cookie` header of the two when using `auth_request` (the first request went fine because the user were communicating with oauth2-proxy not trough `auth_request`). After fixing our NGINX configuration, the issue were gone.

Basically our issue was the same as this: https://github.com/oauth2-proxy/oauth2-proxy/issues/29
The correct snippet we had to apply to our ingress resource is documented here as well: https://oauth2-proxy.github.io/oauth2-proxy/docs/configuration/overview/#configuring-for-use-with-the-nginx-auth_request-directive
